client/auth/auth-code-button.jsx

```diff
@@ -3,65 +3,114 @@
+		this.mounted = false;
+		clearTimeout( this.timeout );
 	}
 
-	refreshData = () => {
-		this.setState( Store.get() );
+	// Reset the notice (which might be in 'error' or 'completed' state) back to 'ready' state.
+	// Happens either when error notice is dismissed, or when 30s timeout elapses.
+	resetSMSCode = () => {
+		clearTimeout( this.timeout );
+		this.timeout = null;
+		this.setState( initialState );
 	};
 
-	requestSMSCode = e => {
-		e.preventDefault();
-		requestCode( this.props.username, this.props.password );
+	requestSMSCode = () => {
+		this.setState( { status: 'requesting' } );
+
+		return request
+			.post( '/sms' )
+			.send( { username: this.props.username, password: this.props.password } )
+			.then(
+				response => this.handleSMSResponse( null, response ),
+				error => this.handleSMSResponse( error, error.response )
+			);
 	};
 
-	render() {
-		const { status, errorLevel, errorMessage } = this.state;
+	handleSMSResponse( error, response ) {
+		if ( ! this.mounted ) {
+			return;
+		}
 
-		let noticeStatus = 'is-info';
-		let showDismiss = false;
-		let message = (
-			<a href="#" onClick={ this.requestSMSCode }>
-				{ this.props.translate( 'Send code via text message.' ) }
-			</a>
-		);
+		this.timeout = setTimeout( this.resetSMSCode, 1000 * 30 );
 
-		if ( status === requestState.REQUESTING ) {
-			message = this.props.translate( 'Requesting code.' );
+		// if it's 2fa error then we actually successfully requested an sms code
+		if ( response && response.body && response.body.error === 'needs_2fa' ) {
+			this.setState( { status: 'complete' } );
+			return;
 		}
 
-		if ( status === requestState.COMPLETE ) {
-			noticeStatus = 'is-success';
-			message = this.props.translate( 'Code sent.' );
+		let errorMessage = null;
+
+		// assign the error message from the response body, otherwise take it from the error object
+		if ( response && response.body && response.body.error_description ) {
+			errorMessage = response.body.error_description;
+		} else if ( error ) {
+			errorMessage = error.message;
 		}
 
-		if ( errorLevel !== false ) {
+		this.setState( { status: 'complete', errorLevel: 'is-error', errorMessage } );
+	}
+
+	render() {
+		const { translate } = this.props;
+		const { status, errorLevel, errorMessage } = this.state;
+
+		let noticeStatus = 'is-info';
+		let showDismiss = false;
+		let message;
+
+		if ( errorLevel ) {
 			noticeStatus = errorLevel;
 			message = errorMessage;
 			showDismiss = true;
+		} else if ( status === 'requesting' ) {
+			message = translate( 'Requesting code.' );
+		} else if ( status === 'complete' ) {
+			noticeStatus = 'is-success';
+			message = translate( 'Code sent.' );
+		} else {
+			/* eslint-disable jsx-a11y/anchor-is-valid */
+			/* eslint-disable jsx-a11y/click-events-have-key-events */
+			/* eslint-disable jsx-a11y/no-static-element-interactions */
+			message = (
+				<a onClick={ this.requestSMSCode }>{ translate( 'Send code via text message.' ) }</a>
```

Comment:
could we flip this to being a text button instead of an anchor and kill all the lint overrides?

Reply:
> could we flip this to being a text button instead of an anchor and kill all the lint overrides?

I'd love to, but that's a material for another UI-only PR. It's not that easy to get all the styling right. Just swapping `a` for `Button` will leave an UI that looks broken.